File: 백준/Silver/1158. 요세푸스 문제/요세푸스 문제.c
#pragma warning(disable:4996)
#include <stdio.h>
#include <stdlib.h>

struct Node {
    struct Node* next;
    struct Node* before;
    int data;
};

struct Node* head = NULL;
struct Node* tail = NULL;
struct Node* makeNode(int data);
void addNode(int data);
void Remove(struct Node* head, int N, int k);

int main() {

    int N; //7
    int K; //3
    int cnt = 0;
    scanf("%d %d", &N, &K);

    for (int i = 1; i <= N; i++)
        addNode(i);

    head->before = tail;
    tail->next = head;

    Remove(head, N, K);

    return 0;

}

struct Node* makeNode(int data) {

    struct Node* new = malloc(sizeof(struct Node));
    new->next = NULL;
    new->before = NULL;
    new->data = data;
    return new;

}

void addNode(int data) {

    struct Node* tmp = head;
    struct Node* newNode = makeNode(data);

    if (head == NULL) {
        head = newNode;
        tail = head;
    }

    else {
        tmp = tail;
        tmp->next = newNode;
        newNode->before = tmp;
        tail = newNode;
    }

}

void Remove(struct Node* head, int N, int K) {

    struct Node* tmp = head->before;
    int i = 0, j = 0;
    printf("<");

    while (i < N) {
        j = K;

        while (j--)
            tmp = tmp->next;

        printf(i++ ? ", %d" : "%d", tmp->data);

        tmp->next->before = tmp->before;
        tmp->before->next = tmp->next;
    }
    printf(">");
}

File: 백준/Silver/1158. 요세푸스 문제/요세푸스 문제.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";
const [N, K] = fs.readFileSync(filePath).toString().split(" ").map(Number);

class Node {
  constructor(data) {
    this.data = data; // 요소의 값
    this.next = null; // 다음 요소를 참조
  }
}

class Queue {
  constructor() {
    this.head = null; // 첫 번째 요소 참조
    this.tail = null; // 마지막 요소 참조
    this.size = 0; // 큐의 길이
  }

  push(data) {
    const newNode = new Node(data);
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail.next = newNode;
      this.tail = newNode;
    }
    this.size++;
  }

  pop() {
    if (!this.head) {
      return null;
    }
    const removeNode = this.head;
    this.head = this.head.next;
    if (!this.head) {
      this.tail = null;
    }
    this.size--;
    return removeNode.data;
  }

  isEmpty() {
    return this.size === 0;
  }
}

const queue = new Queue();
const answer = [];

for (let i = 1; i <= N; i++) {
  queue.push(i);
}

while (!queue.isEmpty()) {

  for (let i = 0; i < K - 1; i++) {
    const front = queue.pop();
    queue.push(front);
  }

  answer.push(queue.pop());
}

console.log(`<${answer.join(", ")}>`);